### ISSUE TITLE
Performance Testing Writeup

### ISSUE THREAD

**Issue by nfaltermeier** [cerner/terra-toolkit#129]
# Google Lighthouse

Scores a webpage in several categories: Performance, Progressive Web App, Accessibility, Best Practices, and Search Engine Optimization. Lighthouse also runs a variety of audits (whether or not the page has or does something) in the above categories. Lighthouse can be configured to only test specific categories or even only run specific audits.

Lighthouse visits a page and tests how long it takes to load, emulating a mobile device by default including network and CPU throttling. Lighthouse doesn't test interaction beyond scrolling down the page.

## Results

We have the option of saving results as a json, html, or csv file. [Lighthouse Report Viewer](https://googlechrome.github.io/lighthouse/viewer/) can take a json report and make a display from it. We also have the option of parsing the report as a javascript object directly if lighthouse is not called directly from the command line. The results show little variance between runs.

Lighthouse also collects traces (see below), logs scripts loaded, logs css rule IDs and whether or not it was used, images loaded, accessibility violations, DOM max width and height, and crawlable links.

## Requirements

* Have Google Chrome installed
* **UPDATE:** Chrome version 66 (April 2018) is required for Lighthouse 3.0.1 (latest), Chrome 54 (October 2016) for Lighthouse 2.9.4 
**[Lighthouse v3 Changelog overview](https://developers.google.com/web/updates/2018/05/lighthouse3)**

## Interesting Links

* **[lightcrawler](https://github.com/github/lightcrawler)** - Crawl a website and run each page found through Lighthouse.
* **[webpack-lighthouse-plugin](https://github.com/addyosmani/webpack-lighthouse-plugin)** - run Lighthouse from a Webpack build.
* **[lighthouse-ci](https://github.com/andreasonny83/lighthouse-ci)** - Run Lighthouse and assert scores satisfy your custom thresholds.
* [The above and other interesting links found here](https://github.com/GoogleChrome/lighthouse#related-projects)

# Chrome Traces

Chrome traces collect an incredibly wide array of information about a page constantly, in the background. A full list of trace categories can be found at [the tracing page](chrome://tracing) by pressing Record then Edit Categories. Traces can be recorded while we run WebdriverIO tests, gathering data for a specific set of actions.

## Results

The results come by default as an incredibly long array of events. There are many processors available to make sense of the events such as the Performance tab in the Dev Tools, the tracing page (chrome://tracing), or even an npm package [devtools-timeline-model](https://github.com/paulirish/devtools-timeline-model).

There is lots of potential information to be gleaned but processing would be dificult. Some potential uses are testing how long a script ran for (confirmed possible) or testing how long it took to download specific resources. [The tracing page](chrome://tracing) can give incredibly detailed information on a trace. Trace results tend to vary from run to run.

## Requirements

* Manipulate webpage with Google Chrome as browser (chromedriver will work)

# Notes

* [Long list of Chrome command line arguments](https://peter.sh/experiments/chromium-command-line-switches/)
* [Notes on using Traces to test Performance](https://gist.github.com/axemclion/079f8bf1a997e9cfe9f0)
* [Selenium capabilities options reference](https://github.com/SeleniumHQ/selenium/wiki/DesiredCapabilities)
* Enabling Performance Traces in WDIO config:

```javascript
capabilities: [
    {
      loggingPrefs: {
        performance: 'ALL',
      },
      'goog:chromeOptions': {
        args: ['--enable-gpu-benchmarking', '--enable-threaded-compositing'],
        perfLoggingPrefs: {
          traceCategories: 'devtools.timeline,disabled-by-default-devtools.timeline,disabled-by-default-devtools.timeline.frame,toplevel',
        },
      },
    },
  ],
```

* Getting performance logs in WebdriverIO test: `const logs = browser.log('performance');`

**Comment by bjankord** [comment]
Nice write-up on the spike. Closing this out.
